packages/ra-core/src/dataProvider/useMutation.ts

```diff
@@ -129,11 +178,131 @@ const useMutation = (
+            JSON.stringify({ query, options }),
+            dataProvider,
+            dataProviderWithDeclarativeSideEffects,
+            setState,
+        ]
+        /* eslint-enable react-hooks/exhaustive-deps */
     );
 
     return [mutate, state];
 };
 
+export interface Mutation {
+    type: string;
+    resource: string;
+    payload: object;
+}
+
+export interface MutationOptions {
+    action?: string;
+    undoable?: boolean;
+    onSuccess?: (response: any) => any | Object;
+    onFailure?: (error?: any) => any | Object;
+    withDeclarativeSideEffectsSupport?: boolean;
+}
+
+export type UseMutationValue = [
+    (query: Partial<Mutation>, options?: Partial<MutationOptions>) => void,
+    {
+        data?: any;
+        total?: number;
+        error?: any;
+        loading: boolean;
+        loaded: boolean;
+    }
+];
+
+/**
+ * Utility function for merging parameters
+ *
+ * useMutation() parameters can be passed:
+ * - at definition time (e.g. useMutation({ type: 'update', resource: 'posts', payload: { id: 1, data: { title: '' } } }) )
+ * - at call time (e.g [mutate] = useMutation(); mutate({ type: 'update', resource: 'posts', payload: { id: 1, data: { title: '' } } }))
+ * - both
+ *
+ * This function merges the definition time and call time parameters.
+ *
+ * This is useful because useMutation() is used by higher-level hooks like
+ * useCreate() or useUpade(), and these hooks can be called both ways.
+ * So it makes sense to make useMutation() capable of handling both call types
+ * as it avoids repetition higher in the hook chain.
+ *
+ * Also, the call time query may be a DOM Event if the callback is used
+ * as an event listener, as in:
+ *
+ * const UpdateButton = () => {
+ *     const mutate = useMutation({ type: 'update', resource: 'posts', payload: { id: 1, data: { title: '' } } });
+ *     return <button onclick={mutate}>Click me</button>
+ * };
+ *
+ * This usage is accepted, and therefore this function checks if the call time
+ * query is an Event, and discards it in that case.
+ *
+ * @param query {Mutation}
+ * @param callTimeQuery {Mutation}
+ * @param options {Object}
+ * @param callTimeOptions {Object}
+ *
+ * @return { type, resource, payload, options } The merged parameters
+ */
+const mergeDefinitionAndCallTimeParameters = (
+    query?: Mutation,
+    callTimeQuery?: Mutation | Event,
+    options?: MutationOptions,
+    callTimeOptions?: MutationOptions
+) => {
+    if (!query && (!callTimeQuery || callTimeQuery instanceof Event)) {
+        throw new Error('Missing query either at definition or at call time');
+    }
+    if (callTimeQuery instanceof Event)
+        return {
+            type: query.type,
+            resource: query.resource,
+            payload: query.payload,
+            options: sanitizeOptions(options),
+        };
+    if (query)
+        return {
+            type: query.type || callTimeQuery.type,
```

Comment:
No, I don't think so: If both `query` and `callTimeQuery` are falsy, the test line 258 will throw an error. So in this line, at least one of the two is of type `Mutation`, which has a required `type`.

Reply:
May be I do not understand something, but we can have `query !== null` and `query.type` falsy. And in this case, since ` callTimeQuery` may be null, a null exception could occur.